### PR TITLE
Adding optional trial argument to model_init

### PR DESCRIPTION
**Pull request by madlag**
# Model structure optimization 

This PR proposes to add the "trial" argument to the model_init function when using `trainer.hyperparameter_search`.
It's backward compatible using six to check the number of arguments of model_init through Python reflection.

```
def model_init(trial):
    if trial != None:
        layer_count = trial.suggest_int("layer_count", 2, 4)
    else:
        layer_count = 2
    return MyModel(layer_count)


trainer = Trainer(
   ...
    model_init=model_init,
   ...
)

trainer.hyperparameter_search(direction="maximize")
```

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -206,6 +206,8 @@ class Trainer:
         model_init (:obj:`Callable[[], PreTrainedModel]`, `optional`):
             A function that instantiates the model to be used. If provided, each call to
             :meth:`~transformers.Trainer.train` will start from a new instance of the model as given by this function.
+            The function may have zero argument, or a single one containing the optuna/Ray Tune trial object, to be able to choose
```
Comment (sgugger) [inline src/transformers/trainer.py:209]:
```suggestion

            The function may have zero argument, or a single one containing the optuna/Ray Tune trial object, to be able to choose
```
nit: sphinx need a blank line to add a new line in the docs.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -206,6 +206,8 @@ class Trainer:
...
+            different architectures according to hyper parameters (such as layer count, sizes of inner layers etc).
```
Comment (sgugger) [inline src/transformers/trainer.py:210]:
```suggestion
            different architectures according to hyper parameters (such as layer count, sizes of inner layers, dropout probabilities etc).
```

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -605,6 +608,19 @@ def _tune_save_checkpoint(self):
                 torch.save(self.optimizer.state_dict(), os.path.join(output_dir, "optimizer.pt"))
                 torch.save(self.lr_scheduler.state_dict(), os.path.join(output_dir, "scheduler.pt"))
 
+    def call_model_init(self, trial = None):
+        import six
```
Comment (sgugger) [inline src/transformers/trainer.py:612]: I'm not familiar with `six`. Is it in the standard library? Any reason not to use the `inspect` module instead?

**Review by sgugger — APPROVED** [review]
Great addition that was a feature requested by some users (to tune dropout values)! Just not familiar with six and I would have used inspect personally, but if it's in the standard library I'm good!
